send/smtp.c

```diff
@@ -576,6 +576,40 @@ static int smtp_auth_oauth(struct SmtpAccountData *adata, const char *method)
   return SMTP_AUTH_SUCCESS;
 }
 
+/**
+ * smtp_auth_xoauth2 - Authenticate an SMTP connection using XOAUTH2
+ * @param adata   SMTP Account data
+ * @param method  Authentication method (not used)
+ * @retval num Result, e.g. #SMTP_AUTH_SUCCESS
+ */
+static int smtp_auth_xoauth2(struct SmtpAccountData *adata, const char *method)
+{
+  (void) method; // This is XOAUTH2
+
+  // L10N: (%s) is the method name, e.g. Anonymous, CRAM-MD5, GSSAPI, SASL
+  mutt_message(_("Authenticating (%s)..."), "XOAUTH2");
+
+  /* We get the access token from the smtp_oauth_refresh_command */
+  char *oauthbearer = mutt_account_getoauthbearer(&adata->conn->account, true);
+  if (!oauthbearer)
+    return SMTP_AUTH_FAIL;
+
+  size_t ilen = strlen(oauthbearer) + 30;
+  char *ibuf = mutt_mem_malloc(ilen);
+  snprintf(ibuf, ilen, "AUTH XOAUTH2 %s\r\n", oauthbearer);
+
+  int rc = mutt_socket_send(adata->conn, ibuf);
+  FREE(&oauthbearer);
+  FREE(&ibuf);
+
+  if (rc == -1)
+    return SMTP_AUTH_FAIL;
+  if (smtp_get_resp(adata) != 0)
+    return SMTP_AUTH_FAIL;
+
+  return SMTP_AUTH_SUCCESS;
+}
```

Comment:
How about refactor `smtp_auth_oauth` and `smtp_auth_xoauth2` into a common function?

```c
static int smtp_auth_oauth_xoauth2(struct SmtpAccountData *adata, const char *method, bool xoauth2)
{
  (void) method; // This is XOAUTH2

  // L10N: (%s) is the method name, e.g. Anonymous, CRAM-MD5, GSSAPI, SASL
  mutt_message(_("Authenticating (%s)..."), method);

  /* We get the access token from the smtp_oauth_refresh_command */
  char *oauthbearer = mutt_account_getoauthbearer(&adata->conn->account, xoauth2);
  if (!oauthbearer)
    return SMTP_AUTH_FAIL;

  size_t ilen = strlen(oauthbearer) + 30;
  char *ibuf = mutt_mem_malloc(ilen);
  snprintf(ibuf, ilen, "AUTH %s %s\r\n", method, oauthbearer);

  int rc = mutt_socket_send(adata->conn, ibuf);
  FREE(&oauthbearer);
  FREE(&ibuf);

  if (rc == -1)
    return SMTP_AUTH_FAIL;
  if (smtp_get_resp(adata) != 0)
    return SMTP_AUTH_FAIL;

  return SMTP_AUTH_SUCCESS;
}

static int smtp_auth_oauth(struct SmtpAccountData *adata, const char *method)
{
  return smtp_auth_oauth_xoauth2(adata, method, false);
}

static int smtp_auth_xoauth2(struct SmtpAccountData *adata, const char *method)
{
  return smtp_auth_oauth_xoauth2(adata, method, true);
}
```

Reply:
Done. Refactored imap authentication as well.